Browserify All The Things!

xinan opened the pull request:
This PR is huge, and kind of scary. :boom: 
Please brace yourself. 

Some notes:
* Remove `requirejs` dependency and related codes. 
* Convert the app from `AMD` to `UMD`.
* Move `editor.html` and `frame.html` from `content/src` to `content`.
* Remove `grunt-replace` as it is no longer needed.
* Remove `servesrc` option from server because it is no longer applicable.
* Use `watchify` for `devserver`, `sdevserver` and `testserver` to auto-recompile.
* Remove `filetype.js` from the list of `grunt-watch` because it will be taken care of by `watchify`.

Feel free to ask questions :smile:

ulularem commented:
It mostly looks good to me (the things I've reviewed and understood look good). I still need to understand the changes to Gruntfile.js and editor-main.js in more detail.

davidbau commented on content/src/FontLoader.js:
```diff
@@ -1,591 +1,589 @@
-    this._numberOfFontFamilies = this._fontFamiliesArray.length;
-    this._fontsMap = {};
-    this._finished = false;
+var isIE = /MSIE/i.test(navigator.userAgent),
```
Even though we're not using define(), can you enclose the FontLoader code in a top-level closure `(function() {/*all the code */}).call(this)` or similar?

davidbau replied:
Also, if you leave the indent level 2-spaces in, the diff will be cleaner.  :-)

davidbau replied:
Oops - I'm new to browserfiy - I guess it probably does this for you.  You can ignore this comment.

dabbler0 commented on Gruntfile.js:
```diff
@@ -333,18 +328,18 @@ module.exports = function(grunt) {
   // "devserver" serves editor code directly from the src directory.
   grunt.registerTask('testserver',
-      ['proxymessage', 'express:localtest', 'watch']);
+      ['proxymessage', 'express:localtest', 'browserify:server', 'watch']);
```
Did you configure `watch` to rebuild the browserify package? Maybe use `watchify` for faster rebuilds if the build is slow.

dabbler0 replied:
How is watchify hooked up to the `devserver`, `sdevserver`, and `testserver` tasks?

xinan replied:
To use `watchify`, simply just set `watch: true` and `keepalive: true` in the `grunt-browserify` options. `devserver`, `sdevserver` and `testserver` use the same `browserify:server` task. Which will watch the source files and do auto-recompile whenever they are modified. 

There is no way to serve source using Browserify. They must be compiled. 

The difference between these 3 servers is that they use different server configs.

davidbau replied:
Ah I see.  https://github.com/jmreidy/grunt-browserify#watch

dabbler0 commented on package.json:
```diff
@@ -30,24 +30,66 @@
     "node-phantom-simple": "latest",
-    "phantomjs": "latest"
+    "phantomjs": "latest",
+    "watchify": "latest"
```
Are you using `watchify` (we should if you're not)?

davidbau commented on package.json:
```diff
@@ -30,24 +30,66 @@
+    "pencil-tracer": "./content/lib/pencil-tracer.js",
+    "see": "./content/lib/see.js",
+    "seedrandom": "./content/lib/seedrandom.js",
+    "sourcemap": "./content/lib/sourcemap/source-map-consumer.js",
```
Can any of these dependencies (such as sourcemap) be brought in via npm instead of bowercopy?

xinan replied:
After some research, seems that `iced-coffee-script`,  `jquery`, `seedrandom` and `source-map` can be brought via `npm`.

davidbau replied:
OK - let's go ahead and switch to using npm for these four dependencies for now.  Also, let's switch droplet - @dabbler0, can you "npm publish" droplet?

davidbau replied:
@xinan - before droplet is npm published, will it work to add a package.json dependency using a github URL - https://docs.npmjs.com/files/package.json#git-urls-as-dependencies ?

sakagg replied:
AFAIK droplet is already pushed to npm as `droplet-editor`

xinan commented:
I've make `uglify` to minimize `editor.js` in the build process to reduce the compiled file from 11MB to 2.2MB. I removed the `pure_getter` option because it is slowing down the minifying process by 4 times and yet the file produced are of almost the same size. 

Also, I removed `browserify:server` task in `testserver` in order to be able to test the minified code.

xinan commented:
Please pull it and run it to do more testings :smile: 

(Run `npm install` first. I updated `npm-shrinkwrap.json`, so tell me if it doesn't work on your machine.)

davidbau commented:
Let's do it!

Merging.
